### PR TITLE
Refinements

### PR DESCRIPTION
**Pull request by MorrisJobke**
* show common HTTP phrases instead of cryptic status codes (404, 401, 503, ...)
* add a lot of log output
* add "requestID" to be able to track logs of a single request and log current version

cc @LukasReschke @nickvergessen

### CODE REVIEW

**Comment by MorrisJobke** [comment]
How to test:

1. setup a local update server that serves an update URL
2. set following two options in `config.php` accordingly, like:
      * `'updater.server.url' => 'http://192.168.0.25/updater_server/',`
      * `'updater.release.channel' => 'daily',`
3. place the index.php of this PR into `updater/index.php`
4. for 9.0.55+, 10.0.1+:
	- apply the patch to have the update button back:
	- `wget https://github.com/nextcloud/server/commit/ad68f22f9b86b878fc0c52d24f6b1ee5f99526be.patch`
	- `patch -p1 < ad68f22f9b86b878fc0c52d24f6b1ee5f99526be.patch`
5. go to admin page, scroll to the updater section and click the "Open updater" button

**Comment by MorrisJobke** [comment]
Also closing the browser should work fine. The only disadvantage then is that you need to generate a new secret and write it to the config.php, because that was done by the "Open updater" button logic and was newer shown to the user.

**Comment by MorrisJobke** [comment]
The last commit addresses https://github.com/nextcloud/server/pull/660 😉

File: index.php [hunk 1]
```diff
@@ -343,8 +362,16 @@ public function checkWritePermissions() {
...
+		if($dir = getenv('NEXTCLOUD_CONFIG_DIR')) {
+			$configFileName = rtrim($dir, '/') . '/config.php';
+		} else {
+			$configFileName = __DIR__ . '/../config/config.php';
+		}
+		$this->silentLog('[info] configFileName ' . $configFileName);
+
 		/** @var array $CONFIG */
-		$configFileName = __DIR__ . '/../config/config.php';
 		require $configFileName;
```
Comment (LukasReschke) [inline index.php]: Add a check if the config file really exists, doesn't have to be the case with `NEXTCLOUD_CONFIG_DIR`.

File: index.php [hunk 2]
```diff
@@ -819,24 +922,125 @@ public function currentStep() {
...
+		if($fh === false) {
+			throw new \LogException('Could not open updater.log');
+		}
+
+		if($this->requestID === null) {
+			$characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
+			$charactersLength = strlen($characters);
+			$randomString = '';
+			for ($i = 0; $i < 10; $i++) {
+				$randomString .= $characters[rand(0, $charactersLength - 1)];
```
Comment (LukasReschke) [inline index.php]: I don't want people complaining about that. Can we just use http://php.net/manual/en/function.random-int.php? 

Also you might want to add `$_SERVER['UNIQUE_ID']` which is used by http://httpd.apache.org/docs/current/mod/mod_unique_id.html

File: index.php [hunk 3]
```diff
@@ -1360,11 +1590,16 @@ function performStep(number, callback) {
...
-						text += '<li>' + file + '</li>';
-					});
-					text += '</ul>';
+					var text = '';
+					if (typeof response['response'] === 'string') {
+						text = response['response'];
+					} else {
+						text = 'The following extra files have been found:<ul>';
+						response['response'].forEach(function(file) {
+							text += '<li>' + file + '</li>';
```
Comment (LukasReschke) [inline index.php]: Also worth escaping.

File: index.php [hunk 4]
```diff
@@ -1360,11 +1590,16 @@ function performStep(number, callback) {
...
 					errorStep('step-check-files');
 
-					var text = 'The following extra files have been found:<ul>';
-					response['response'].forEach(function(file) {
-						text += '<li>' + file + '</li>';
-					});
-					text += '</ul>';
+					var text = '';
+					if (typeof response['response'] === 'string') {
+						text = response['response'];
```
Comment (LukasReschke) [inline index.php]: Let's escape that.

File: index.php [hunk 5]
```diff
@@ -1376,11 +1611,16 @@ function performStep(number, callback) {
...
+					} else {
+						text = 'The following places can not be written to:<ul>';
+						response['response'].forEach(function(file) {
+							text += '<li>' + file + '</li>';
```
Comment (LukasReschke) [inline index.php]: Should be escaped.

File: index.php [hunk 6]
```diff
@@ -1376,11 +1611,16 @@ function performStep(number, callback) {
...
 					errorStep('step-check-permissions');
 
-					var text = 'The following places can not be written to:<ul>';
-					response['response'].forEach(function(file) {
-						text += '<li>' + file + '</li>';
-					});
-					text += '</ul>';
+					var text = '';
+					if (typeof response['response'] === 'string') {
+						text = response['response'];
```
Comment (LukasReschke) [inline index.php]: Should be escaped.

File: index.php (hunk 2 above)
Comment (MorrisJobke) [inline index.php]: It's for making the log more readable.

**Comment by MorrisJobke** [comment]
@LukasReschke I addressed your comments

**Comment by LukasReschke** [comment]
Thanks, @MorrisJobke :)

**Comment by MorrisJobke** [comment]
Together with https://github.com/nextcloud/server/pull/1499 (and it's backports) this is what we need for our updater to work flawlessly - cc @nickvergessen @LukasReschke

**Comment by nickvergessen** [comment]
Tested:
1. Have additional files in place
2. Start update
3. See error that additional files exist
4. Delete files
5. Restart the update with the "Start update" button

**Expected:** List of additional files should be empty

Same "error" happens with "check write permissions".
The list should be cleared on a rerun

**Comment by nickvergessen** [comment]
Other then that it worked fine.

**Comment by nickvergessen** [comment]
👍

**Comment by MorrisJobke** [comment]
> 5. Restart the update with the "Start update" button

Good point. This should usually not be possible or otherwise better clear the current state of the UI.

**Comment by MorrisJobke** [comment]
> Same "error" happens with "check write permissions".
> The list should be cleared on a rerun

Fixed.

> 5. Restart the update with the "Start update" button

I fixed this properly. There is now a "Retry" button, that removes the warning of the last step and reruns it. The buttons are only shown, when they are actually needed.

**Comment by MorrisJobke** [comment]
@LukasReschke @nickvergessen Could you have a look at the latest changes? thanks

**Comment by MorrisJobke** [comment]
I just tested it with an 9.0.53 and this updater on IE11 and it works there fine too.

**Comment by MorrisJobke** [comment]
🎉

**Comment by LukasReschke** [comment]
RCs with the updater are built and available at:

=> https://download.nextcloud.com/server/prereleases/nextcloud-10.0.1RC2.tar.bz2
=> https://download.nextcloud.com/server/prereleases/nextcloud-9.0.54RC2.tar.bz2

**Comment by blizzz** [comment]
Upgraded my instance to the 10.0.1RC2, worked fine.

**Pull request merged**